### PR TITLE
JSAAL-15 Finished SignIn Function and Login css style

### PR DESCRIPTION
**Pull request by Aaron-Xiaoxingchen**
pleas help review code.Thanks

### CODE REVIEW

File: src/Account/LogIn.jsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  const { accessToken } = useSelector((state) => state.users);
+
+  useEffect(() => {
+    if (accessToken) pop();
+  }, [accessToken]);
+
+  // eslint-disable-next-line consistent-return
+  const signIn = () => {
+    if (username === '' || password === '') {
+      return Alert.alert('All the Input Field are required');
```
Comment (FairyWind) [inline src/Account/LogIn.jsx]: Pay attention to word case

File: src/Account/index.jsx [hunk 2]
```diff
@@ -1,9 +1,42 @@
...
-    <Text>This is the account</Text>
-  </View>
-);
+import { View, Text, Button } from 'react-native';
+import { useSelector, useDispatch } from 'react-redux';
+import { useNavigation } from 'react-navigation-hooks';
+import { logOut } from '../redux/account/actionCreator';
+import styles from './styles';
+
+
```
Comment (FairyWind) [inline src/Account/index.jsx]: More blank lines

File: src/redux/account/actionCreator.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import {
+  LOGIN_USER_FAILURE,
+  LOGIN_USER_SUCCESS,
+  LOGOUT_SUCCESS,
+} from './actionType';
+
+const loginUser = (username, password) => (dispatch) => {
+  const LOCAL_URL = 'http://10.72.161.57:3000';
```
Comment (FairyWind) [inline src/redux/account/actionCreator.js]: move this to .env

File: src/store.js [hunk 4]
```diff
@@ -6,17 +6,20 @@ import thunk from 'redux-thunk';
...
   search: searchReducer,
   watchlists: watchlistsReducer,
+  users: accountReducer,
 });
 
 const persistConfig = {
   key: 'root',
   storage: AsyncStorage,
   whitelist: ['watchlists'],
+  blacklist: ['users'],
```
Comment (FairyWind) [inline src/store.js]: delete this test line

File: src/redux/account/actionCreator.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+  })
+    .then((response) => {
+      if (response.status === 200) {
+        return response.json();
+      }
+      throw new Error('Unexpected status code');
+    })
+    .then((response) => {
+      dispatch({ type: LOGIN_USER_SUCCESS, payload: { username, ...response } });
+    })
```
Comment (FairyWind) [inline src/redux/account/actionCreator.js]: create actions for success and fail, don't dispatch obejct here

File: src/Account/styles.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+  container: {
+    flex: 1,
+    alignItems: 'center',
+    justifyContent: 'center',
+    flexDirection: 'column',
+  },
+  textInput: {
+    borderRadius: 5,
+    backgroundColor: '#fff',
+    // width: 300,
```
Comment (FairyWind) [inline src/Account/styles.js]: delete this comment line

**Review by FairyWind — CHANGES_REQUESTED** [review]
please fix problems

File: src/redux/account/actionCreator.js [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    body: JSON.stringify({
+      username,
+      password,
+    }),
+  })
+    .then((response) => {
+      if (response.status === 200) {
+        return response.json();
+      }
+      throw new Error('Unexpected status code');
```
Comment (FairyWind) [inline src/redux/account/actionCreator.js]: The description of the error is vague

File: src/redux/account/actionCreator.js [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      dispatch(loginUserSuccess(response, username));
+    })
+    .catch((error) => dispatch(loginUserFail(error)));
+};
+
+export const logOut = () => ({
+  type: LOGOUT_SUCCESS,
+});
+
+
```
Comment (FairyWind) [inline src/redux/account/actionCreator.js]: delete more blank

**Review by FairyWind — CHANGES_REQUESTED** [review]
still have some problems

**Review by FairyWind — APPROVED** [review]

File: .env.template [hunk 9]
```diff
@@ -1,2 +1,3 @@
 API_URL=https://sandbox.iexapis.com/stable
 API_KEY=abcdefg1234567890
+LOCAL_URL=http://10.72.161.57:3000;
```
Comment (ajaxjiang96) [inline .env.template]: Please use another name, `LOCAL_URL` does not make sense (we will host our server sometime in the future, and that's not local)

File: src/Account/index.jsx [hunk 10]
```diff
@@ -1,9 +1,41 @@
...
+          <View>
+            <Text style={{ fontSize: 20 }}>
+              Welcome to our app
+              {' '}
+              {username}
+            </Text>
+            <Button style={styles.signOutButton} title="Logout" onPress={handleLogOut} />
+          </View>
+        )
+        : (
```
Comment (ajaxjiang96) [inline src/Account/index.jsx]: `) : (`

File: src/navigation/MainTabNavigation.jsx [hunk 11]
```diff
@@ -7,6 +7,7 @@ import Watchlists from '../Watchlists';
 import Investments from '../Investments';
 import Account from '../Account';
 import Playground from '../Playground';
+import Login from '../Account/LogIn';
```
Comment (ajaxjiang96) [inline src/navigation/MainTabNavigation.jsx]: `Login`

**Review by ajaxjiang96 — DISMISSED** [review]

File: src/Account/LogIn.jsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import Row from '../common/Row';
+import Column from '../common/Column';
+
+const SignInButton = ({ onPress }) => (
+  <TouchableHighlight style={styles.button} onPress={onPress} underlayColor="#5d70ba" activeOpacity={0.5}>
+    <Text style={styles.buttonText}><FontAwesome5 name="arrow-right" size={16} /></Text>
+  </TouchableHighlight>
+);
+
+const SignIn = () => {
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: I know signin and login are the same, but we should use them consequently. Now the filename and the component name is not matching. Pls forget the SignIn and use `Login` everywhere. Like LoginButton, Login.jsx etc...

File: src/Account/LogIn.jsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  const [username, setUsername] = useState('');
+  const [password, setPassword] = useState('');
+  const dispatch = useDispatch();
+  const { pop } = useNavigation();
+
+  const { accessToken } = useSelector((state) => state.users);
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: do we want to handle more users at the same time? It should be simply `state.user`, not `state.users`

File: src/Account/LogIn.jsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+  useEffect(() => {
+    if (accessToken) pop();
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: whats going on here?

File: src/Account/index.jsx [hunk 15]
```diff
@@ -1,9 +1,41 @@
...
+    <View style={styles.container}>
+      {accessToken
+        ? (
+          <View>
+            <Text style={{ fontSize: 20 }}>
+              Welcome to our app
+              {' '}
+              {username}
+            </Text>
+            <Button style={styles.signOutButton} title="Logout" onPress={handleLogOut} />
```
Comment (ikarasz) [inline src/Account/index.jsx]: so is it LogOut or Logout? :smile_cat:  Let's go with logout

File: src/Account/styles.js [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  },
+  loginTitle: {
+    fontSize: 36,
+    fontWeight: 'bold',
+  },
+  inputContainer: {
+    flex: 1,
+    padding: 15,
+  },
+  signOutButton: {
```
Comment (ikarasz) [inline src/Account/styles.js]: signout or logout?

File: src/Watchlists/WatchList.jsx [hunk 17]
```diff
@@ -95,7 +95,7 @@ export default function Watchlist({ item }) {
       {isOpen && (
         <>
           {isEmptyWatchList
-            ? <View style={styles.editorText}><Text style={{ color: '#ccc' }}>This WatchList is Empty</Text></View>
+            ? <View style={styles.editorText}><Text style={{ color: '#ccc' }}>This watchList is empty</Text></View>
```
Comment (ikarasz) [inline src/Watchlists/WatchList.jsx]: Please avoid unrelated changes. This should be another pullrequest

File: src/redux/account/actionType.js [hunk 18]
```diff
@@ -0,0 +1,3 @@
+export const LOGIN_USER_SUCCESS = 'LOGIN_USER_SUCCESS';
```
Comment (ikarasz) [inline src/redux/account/actionType.js:2]: Don't we need a start action as well?

File: src/redux/account/reducer.js [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import {
+  LOGIN_USER_SUCCESS,
+  LOGIN_USER_FAILURE,
+  LOGOUT_SUCCESS,
+} from './actionType';
+
+export default (state = {}, action) => {
```
Comment (ikarasz) [inline src/redux/account/reducer.js]: please define an initial state

File: src/redux/account/actionCreator.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+    body: JSON.stringify({
+      username,
+      password,
+    }),
+  })
+    .then((response) => {
+      if (response.status === 200) {
+        return response.json();
+      }
+      throw new Error('Oops, there\'s something wrong with our app.');
```
Comment (ikarasz) [inline src/redux/account/actionCreator.js]: and what happens if the password was wrong?

**Review by ikarasz — CHANGES_REQUESTED** [review]

File: src/Account/LogIn.jsx (hunk 14 above)
Comment (Aaron-Xiaoxingchen) [reply src/Account/LogIn.jsx]: login Button and logout button are in the same page called index.jsx and index.jsx and login.jsx are in the same stackNavigator. when get token pop the login.jsx page.

File: src/Watchlists/WatchList.jsx (hunk 17 above)
Comment (Aaron-Xiaoxingchen) [reply src/Watchlists/WatchList.jsx]: Okay, sorry!

**Review by ikarasz — CHANGES_REQUESTED** [review]
please ignore the unnecessary files

File: src/Account/LogIn.jsx [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+import Row from '../common/Row';
+import Column from '../common/Column';
+
+const SignInButton = ({ onPress }) => (
+  <TouchableHighlight style={styles.button} onPress={onPress} underlayColor="#5d70ba" activeOpacity={0.5}>
+    <Text style={styles.buttonText}><FontAwesome5 name="arrow-right" size={16} /></Text>
+  </TouchableHighlight>
+);
+
+const LogIn = () => {
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: 10 lines above it is SignIn and here it is LogIn, pls be consequent use only one expression for the same action

File: src/Account/LogIn.jsx [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+
+  useEffect(() => {
+    if (accessToken) pop();
+  }, [accessToken]);
+
+  const logIn = () => {
+    if (username === '' || password === '') {
+      return Alert.alert('All the input field are required');
+    }
+    return dispatch(loginUser(username, password));
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: the action is login the component is logIn, pls decide if it is one word or two and then be consequent

File: src/Account/LogIn.jsx [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+
+  const { accessToken } = useSelector((state) => state.user);
+
+  useEffect(() => {
+    if (accessToken) pop();
+  }, [accessToken]);
+
+  const logIn = () => {
+    if (username === '' || password === '') {
+      return Alert.alert('All the input field are required');
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: do not use Alert for error messages, try to common ErrorMessage component instead

File: src/Account/index.jsx [hunk 24]
```diff
@@ -1,9 +1,41 @@
...
+          <View>
+            <Text style={{ fontSize: 20 }}>
+              Welcome to our app
+              {' '}
+              {username}
+            </Text>
+            <Button style={styles.logOutButton} title="logout" onPress={handleLogOut} />
+          </View>
+        )
+        : (
```
Comment (ikarasz) [inline src/Account/index.jsx]: pls apply this change

File: src/redux/account/actionCreator.js [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+});
+
+export const loginUserFail = (payload) => ({
+  type: LOGIN_USER_FAILURE,
+  payload,
+});
+
+const loginUser = (username, password) => (dispatch) => {
+  const loginURL = `${BACKEND_URL}/users/login`;
+  fetch(loginUserStart);
```
Comment (ikarasz) [inline src/redux/account/actionCreator.js]: what is this?

File: src/redux/account/actionCreator.js [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+    body: JSON.stringify({
+      username,
+      password,
+    }),
+  })
+    .then((response) => {
+      if (response.status === 200) {
+        return response.json();
+      } if (response.status === 401) {
+        Alert.alert('Your username or password is incorrect');
```
Comment (ikarasz) [inline src/redux/account/actionCreator.js]: do not do any UI stuff in the action, lets dipatch en error and handle that in the corresponding component

File: src/redux/account/reducer.js [hunk 27]
```diff
@@ -0,0 +1,37 @@
+import {
+  LOGIN_USER_START,
+  LOGIN_USER_SUCCESS,
+  LOGIN_USER_FAILURE,
+  LOGOUT_SUCCESS,
+} from './actionType';
+
+const initialState = {
+  accessToken: '',
+  error: '',
```
Comment (ikarasz) [inline src/redux/account/reducer.js:10]: pls add some isLogginIn state, so you can handle that in the component. Otherwise the user wont know whats going on if the log in takes too long time (eg on slow network)

**Review by ikarasz — CHANGES_REQUESTED** [review]

File: src/Account/index.jsx (hunk 24 above)
Comment (Aaron-Xiaoxingchen) [reply src/Account/index.jsx]: what does this mean?

File: src/redux/account/actionCreator.js (hunk 25 above)
Comment (Aaron-Xiaoxingchen) [reply src/redux/account/actionCreator.js]: can I apply this for loading?

File: src/redux/account/reducer.js [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+  isLogginIn: false,
```
Comment (ikarasz) [inline src/redux/account/reducer.js]: it should be `isLoggingIn`

File: src/redux/account/reducer.js [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+};
+
+export default (state = initialState, action) => {
+  switch (action.type) {
+    case LOGIN_USER_START:
+      return {
+        ...state,
```
Comment (ikarasz) [inline src/redux/account/reducer.js:18]: don't we need to set `isLoggingIn` to true?

File: src/redux/account/reducer.js [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+      };
+    case LOGIN_USER_SUCCESS:
+      return {
+        ...state,
+        accessToken: action.payload,
+        isLogginIn: true,
```
Comment (ikarasz) [inline src/redux/account/reducer.js]: when the user logged in successfully then `isLoggingIn` should be false, since the process has ended.

File: src/Account/index.jsx [hunk 31]
```diff
@@ -1,9 +1,38 @@
...
+          <View>
+            <Text style={{ fontSize: 20 }}>
+              {`Welcome to our app, ${username}`}
+            </Text>
+            <Button style={styles.logOutButton} title="logout" onPress={handleLogOut} />
+          </View>
+        ) : (
+          <View>
+            <Text>Please Login</Text>
+            <Button title="Login" onPress={() => push('Login')} />
```
Comment (ikarasz) [inline src/Account/index.jsx]: please be consequent and use either capitalized titles or  lowercased, but do not mix them
```suggestion
            <Button title="login" onPress={() => push('Login')} />
```

**Review by ikarasz — CHANGES_REQUESTED** [review]

File: src/Account/LogIn.jsx [hunk 32]
```diff
@@ -0,0 +1,93 @@
+import React, { useState, useEffect } from 'react';
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: why is this file `LogIn.jsx` again? It was `Login.jsx` in the last pr :rofl:

File: src/Account/index.jsx (hunk 24 above)
Comment (ikarasz) [reply src/Account/index.jsx]: it means that please replace
```
)
: (
```
with
```
) : (
```
As @ajaxjiang96 suggested

File: src/redux/account/reducer.js [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+    case LOGIN_USER_START:
+      return {
+        ...state,
+        isLoggingIn: true,
+      };
+    case LOGIN_USER_SUCCESS:
+      return {
+        ...state,
+        accessToken: action.payload,
+        isLogginIn: false,
```
Comment (ikarasz) [inline src/redux/account/reducer.js]: did you try your code? This shouldn't work
```suggestion
        isLoggingIn: false,
```

File: src/Account/LogIn.jsx [hunk 34]
```diff
@@ -0,0 +1,93 @@
...
+
+  useEffect(() => {
+    if (accessToken) pop();
+  }, [accessToken]);
+
+  const login = () => {
+    dispatch(loginUser(username, password));
+  };
+
+  return (
```
Comment (ikarasz) [inline src/Account/LogIn.jsx]: I've just realized that you didn't add any loader for the login sad :(
Please use the isLoggingIn state, otherwise it has no point to have it...

File: src/redux/account/reducer.js [hunk 35]
```diff
@@ -0,0 +1,43 @@
...
+      };
+    case LOGIN_USER_FAILURE:
+      return {
+        ...state,
+        error: action.payload,
+        isLogginIn: false,
```
Comment (ikarasz) [inline src/redux/account/reducer.js]:
```suggestion
        isLoggingIn: false,
```

File: src/redux/account/reducer.js [hunk 36]
```diff
@@ -0,0 +1,43 @@
...
+      };
+    case LOGOUT_SUCCESS:
+      return {
+        ...state,
+        accessToken: null,
+        isLogginIn: false,
```
Comment (ikarasz) [inline src/redux/account/reducer.js]:
```suggestion
        isLoggingIn: false,
```

**Review by ikarasz — CHANGES_REQUESTED** [review]

**Review by ikarasz — APPROVED** [review]
